### ISSUE TITLE
scale

### ISSUE THREAD

**Issue by foralliance** [zisianw/FaceBoxes.PyTorch#6]
@zisianw @sfzhang15  HI
关于C.Relu模块，在**Fig2(a)**中，有1个**scale层**．
但在code并没有体现出该层．
请问如何理解？？

**Comment by sfzhang15** [comment]
@foralliance 
caffe里BN层和Scale层是分开的，PyTorch是在一块的，所以这里稍微有点不一样

**Comment by foralliance** [comment]
谢谢！

另外一个问题：
code中，wider数据集是按照voc格式制作的．在voc的xml格式中，对每一个box有一个**参数difficult**．对于wider数据集，这个diffcult的取值您是如何设置的？什么样的box才是difficult呢？

当然，在`AnnotationTransform函数`中，因为`keep_difficult=True`的设置，保证了判断
`if not self.keep_difficult and difficult`
始终为**false**．所以不论box的**difficult值**如何设置．所有的box都会参与后面的计算．

**Comment by sfzhang15** [comment]
@foralliance 
WIDER FACE中有一个invalid的标签，invalid=1的都可以置为difficult=true并在训练中忽略。你可以试试

**Comment by foralliance** [comment]
@sfzhang15 
确实如此．谢谢

另外一个问题：
在FaceBoxes的**caffe代码**中，conv1和conv2后的2个Pooling层的pad=0（[models/faceboxes/train.prototxt](https://github.com/sfzhang15/FaceBoxes/blob/master/models/faceboxes/train.prototxt)）.
但是在**pytorch代码**中，conv1和conv2后的2个Pooling层的pad却是1（[model/faceboxes.py中的forward部分](https://github.com/zisianw/FaceBoxes.PyTorch/blob/master/models/faceboxes.py)）：
` x = F.max_pool2d(x, kernel_size=3, stride=2, padding=1)`

不知道是有意为之还是？？

**Comment by zisianw** [comment]
Hi @foralliance, thanks for noticing! That was indeed a negligence in the PyTorch implementation as it turns out that pooling layer in [Caffe](https://github.com/sfzhang15/FaceBoxes/blob/master/src/caffe/layers/pooling_layer.cpp#L90) uses ceil and [PyTorch](https://pytorch.org/docs/stable/nn#maxpool2d) uses floor by default.
To have an identical implementation on this part, it should have been:
`x = F.max_pool2d(x, kernel_size=3, stride=2, padding=0, ceil_mode=True)`

**Comment by foralliance** [comment]
@zisianw 
got it!!
thanks

**Comment by foralliance** [comment]
@zisianw  HI
一个额外的问题麻烦下：
在DataLoader中，有个参数num_workers，发现这个参数严重影响着数据读取速度．
```
Epoch:1 || epochiter: 0.0/402.0|| Totel iter 0 || L: 9.9270 C: 23.2169||Batch time: 13.3025 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 1.0/402.0|| Totel iter 1 || L: 9.5204 C: 20.6763||Batch time: 0.8111 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 2.0/402.0|| Totel iter 2 || L: 9.7094 C: 14.3963||Batch time: 0.8572 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 3.0/402.0|| Totel iter 3 || L: 9.2358 C: 12.4466||Batch time: 0.7305 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 4.0/402.0|| Totel iter 4 || L: 8.7333 C: 8.5753||Batch time: 0.6582 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 5.0/402.0|| Totel iter 5 || L: 9.3984 C: 7.7155||Batch time: 0.6882 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 6.0/402.0|| Totel iter 6 || L: 8.3960 C: 6.0018||Batch time: 0.7805 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 7.0/402.0|| Totel iter 7 || L: 9.2684 C: 6.2540||Batch time: 0.6133 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 8.0/402.0|| Totel iter 8 || L: 9.3614 C: 5.6496||Batch time: 6.2326 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 9.0/402.0|| Totel iter 9 || L: 8.6283 C: 5.4357||Batch time: 0.7247 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 10.0/402.0|| Totel iter 10 || L: 8.4319 C: 4.9669||Batch time: 1.4879 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 11.0/402.0|| Totel iter 11 || L: 8.4364 C: 4.1630||Batch time: 0.6911 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 12.0/402.0|| Totel iter 12 || L: 8.4593 C: 4.6689||Batch time: 0.5316 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 13.0/402.0|| Totel iter 13 || L: 8.7163 C: 4.6978||Batch time: 0.6517 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 14.0/402.0|| Totel iter 14 || L: 8.6776 C: 4.4363||Batch time: 0.7663 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 15.0/402.0|| Totel iter 15 || L: 8.2896 C: 4.4251||Batch time: 0.6142 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 16.0/402.0|| Totel iter 16 || L: 8.6511 C: 4.2516||Batch time: 8.1611 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 17.0/402.0|| Totel iter 17 || L: 8.6909 C: 4.2532||Batch time: 0.5894 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 18.0/402.0|| Totel iter 18 || L: 8.6228 C: 4.1899||Batch time: 0.5962 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 19.0/402.0|| Totel iter 19 || L: 8.3660 C: 5.0475||Batch time: 0.7253 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 20.0/402.0|| Totel iter 20 || L: 8.3578 C: 3.9863||Batch time: 0.6487 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 21.0/402.0|| Totel iter 21 || L: 8.0489 C: 3.8586||Batch time: 0.6625 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 22.0/402.0|| Totel iter 22 || L: 7.7491 C: 4.0214||Batch time: 0.6410 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 23.0/402.0|| Totel iter 23 || L: 8.2842 C: 3.7695||Batch time: 0.8866 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 24.0/402.0|| Totel iter 24 || L: 8.5159 C: 3.8839||Batch time: 6.0781 sec. ||LR: 0.00100000
Epoch:1 || epochiter: 25.0/402.0|| Totel iter 25 || L: 8.3251 C: 3.8723||Batch time: 0.6119 sec. ||LR: 0.00100000
```
配置：单GPU(1080ti)，batch_size=32，num_workers=8．
发现：每迭代8的倍数，当前迭代时间就会增加尽10倍．
自己的理解是num_workers=8，应该就是一次8个线程处理数据．所以每到8的倍数，要开始处理，所以时间增加．
**不知道这样的理解对否？？**

修改num_workers为其他数值，也发现同样的规律：只要当前迭代数是num_workers值的倍数，时间都会增加尽10倍．
**这个问题有什么好的解决办法吗？？**
另外修改num_workers值，会影响数据的读取顺序，**那么理论上也会对最终的结果有影响吧？**
